test/test_dynamic_transform.cpp

```diff
@@ -0,0 +1,485 @@
+  auto tv3 = add(tv1, tv2);
+
+  fusion.addOutput(tv3);
+
+  fusion.printMath();
+
+  // tv2 has symbolic axes as reshape is dynamic
+  TORCH_CHECK(
+      tv2->domain()->hasSymbolicAxis(),
+      "Expected to have symbolic axes: ",
+      tv2->toString());
+
+  // The symbolic axes of tv2 should not be propagated to tv3 as tv1
+  // is fully concrete
+  TORCH_CHECK(
+      !tv3->domain()->hasSymbolicAxis(),
+      "Not expected to have symbolic axes: ",
+      tv3->toString());
+
+  {
+    ExpressionEvaluator expr_eval;
+
+    // input: 4, 3
+    // output: 3, 4
+    expr_eval.bind(tv0->axis(0)->extent(), 4);
+    expr_eval.bind(tv0->axis(1)->extent(), 3);
+    expr_eval.bind(reshape_shape0, 3);
+    expr_eval.bind(reshape_shape1, 4);
+
+    auto info = DynamicTransform::getConcretizationInfo(&fusion, &expr_eval);
+
+    std::cerr << info.toString() << std::endl;
+  }
+
+  {
+    ExpressionEvaluator expr_eval;
+
+    // input: 4, 3
+    // output: 3, -1
+    expr_eval.bind(tv0->axis(0)->extent(), 4);
+    expr_eval.bind(tv0->axis(1)->extent(), 3);
+    expr_eval.bind(reshape_shape0, 3);
+    expr_eval.bind(reshape_shape1, -1);
+
+    auto info = DynamicTransform::getConcretizationInfo(&fusion, &expr_eval);
+
+    std::cerr << info.toString() << std::endl;
+  }
+
+  {
+    ExpressionEvaluator expr_eval;
+
+    // input: 4, 3
+    // output: 5, -1
+    expr_eval.bind(tv0->axis(0)->extent(), 4);
+    expr_eval.bind(tv0->axis(1)->extent(), 3);
+    expr_eval.bind(reshape_shape0, 5);
+    expr_eval.bind(reshape_shape1, -1);
+
+    // This should fail as (4 * 3) is not evenly divisible by 5
+    EXPECT_THAT(
+        [&]() { DynamicTransform::getConcretizationInfo(&fusion, &expr_eval); },
+        ::testing::ThrowsMessage<c10::Error>(
+            ::testing::HasSubstr("Cannot infer")));
+  }
+}
+
+TEST_F(NVFuserTest, DynamicTransform2_CUDA) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  // All tensors are 2D symbolic tensors. tv1 and tv2 have the same shape
+
+  auto tv0 = makeSymbolicTensor(2);
+  fusion.addInput(tv0);
+  auto tv1 = makeSymbolicTensor(2);
+  fusion.addInput(tv1);
+  auto tv2 = makeSymbolicTensor(2);
+  fusion.addInput(tv2);
+
+  // Reshape to the same shape as tv1
+  auto tv3 = reshape(tv0, {tv1->axis(0)->extent(), tv1->axis(1)->extent()});
+  auto tv4 = add(tv1, tv2);
+  auto tv5 = add(tv3, tv4);
+  fusion.addOutput(tv5);
+
+  fusion.printMath();
+
+  {
+    ExpressionEvaluator expr_eval;
+
+    // input: 4, 3
+    // output: 3, 4
+    expr_eval.bind(tv0->axis(0)->extent(), 4);
+    expr_eval.bind(tv0->axis(1)->extent(), 3);
+    // Bind only tv2 extents. It should be enough as tv1 has the same
+    // shape
+    expr_eval.bind(tv2->axis(0)->extent(), 3);
+    expr_eval.bind(tv2->axis(1)->extent(), 4);
+
+    auto info = DynamicTransform::getConcretizationInfo(&fusion, &expr_eval);
+
+    std::cerr << info.toString() << std::endl;
+  }
+}
+
+TEST_F(NVFuserTest, DynamicTransform3_CUDA) {
```

Comment:
@jjsjann123 This is a simple example of defining a symbolic reshape and later concretizing it with actual sizes.

Line 172 is where the symbolic fusion is analyzed with actual sizes. Line 176 is where the fusion is concretized, i.e., the symbolic reshape is converted to a static reshape. Thereafter, the fusion can be fed to the rest of the system, i.e., segmenter and GpuLower.